### ISSUE TITLE
Upgrade v0.20.0 and dockerenable = false

### ISSUE THREAD

**Issue by vinvli** [actions-runner-controller/actions-runner-controller#835]
Deployed the new v0.20.0 version but if I add the flag dockerEnabled: false it fails with the error in the controller log:

`Reconciler error	{"reconciler group": "actions.summerwind.dev", "reconciler kind": "Runner", "name": "github-runner-bn4xw-55gx9", "namespace": "github-runner", "error": "Pod \"github-runner-bn4xw-55gx9\" is invalid: spec.containers[1].image: Required value"}`

We use the following example deployment:

```
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: github-runner
  namespace: github-runner
spec:
  replicas: 1
  template:
    spec:
      organization: org-IT
      ephemeral: true
      dockerEnabled: false
      resources:
        limits:
          memory: "2Gi"
        requests:
          memory: "1Gi"
      image: privateacr.azurecr.io/github/actions-runner:v2.283.1-ubuntu-20.04
      env:
      - name: NODE_EXTRA_CA_CERTS
        value: /usr/local/share/ca-certificates/rootca.pem
```

If I revert back to 0.19.0 it works as expected (no dind is injected)

**Comment by tyrken** [comment]
FWIW this seems the same issue that has a little discussion at the end (scroll down) of #801

**Comment by tyrken** [comment]
@mumoshu - it seems https://github.com/actions-runner-controller/actions-runner-controller/blob/24602ff3ee05303e019b395b5d6805f2ebedaa14/controllers/runner_controller.go#L569 has the responsibility for taking the runner spec - containing a bunch of keys like image, dockerEnabled, VolumeMounts, etc. and creating a `corev1.Pod`, to be either used directly making a pod or embedding in another resource as a PodSpec.

Rather than making the CRD more flexible in accepting unknown/preserved fields, can't we correct the logic of this function to make a valid pod?  I.e. not appending the second "docker" pod when dockerEnabled is false?

**Comment by mumoshu** [comment]
> Rather than making the CRD more flexible in accepting unknown/preserved fields, can't we correct the logic of this function to make a valid pod? 

Nope, because AFAIK it's k8s-apiserver rather than our controller that is emitting validation errors.

**Comment by aliusmiles** [comment]
As per advise from #801 the following config fixed this issue for me:

```yaml
  dockerEnabled: false
  dockerdWithinRunnerContainer: true
```

**Comment by tyrken** [comment]
@mumoshu I'm not talking about stopping the validation in K8s apiserver (as removing the CRD section might do), but fixing the bad data it's reasonably complaining about.  Please look at PR #837 which lets RunnerDeployments with dockerEnabled: false succeed.

**Comment by mumoshu** [comment]
@tyrken Ah thanks! Probably I have conflated two different issues. Yours seems to fix one of them. Let me review it in a few days.

**Comment by mumoshu** [comment]
Thanks everyone- I now understand https://github.com/actions-runner-controller/actions-runner-controller/issues/801#issuecomment-923155679 by @yuri-1987 was a dupe of this one. This will be fixed via #837.
